### PR TITLE
Fix issues related to listings

### LINKED ISSUE #236: Issues related to listings
### Description

I experienced several issues related to listings, such as:
- the wide listing (`listing*`) environment is not available (should eventually be fixed in #225),
- the caption and the listing are overlapping in the `marginlisting` environment (as noted in Sect. 3.5 of the documentation: "Unfortunately, the space between the caption and the listing must be adjusted manually; if you find a better way, please let me know."),
- the captions of the `lstlisting` environment are counted within chapter, but the captions of the `listing` environment are not (also they are counted independently, `marginlisting` uses the former, and thus currently there is no way to make the counter of `marginlisting` consistent with minted), and
- hyphens are printed before and after each margin listing (not only in listings but also in figures and tables, fixed in #235).

I'm preparing a PR to fix (most of) them. I provide a test case for my fix below.

### Minimal Working Example

```latex
\documentclass{kaobook}[twoside=true]
\usepackage[english]{babel}
\usepackage{lipsum}
\usepackage{kaobiblio}
\usepackage{minted}
\usepackage{listings}

\begin{document}

\setchapterstyle{kao}

\chapter{First Chapter}

\begin{listing}[h]
\caption{Hello world in C (minted, normal listing).}
\begin{minted}{c}
#include <stdio.h>
int main() {
  printf("Hello, World!");
  return 0;
}
\end{minted}
\end{listing}

\lipsum[1]

\begin{listing*}[h]
\caption{Hello world in C (minted, wide listing).}
\begin{minted}{c}
#include <stdio.h>
int main() { printf("Hello, World!"); return 0; }
\end{minted}
\end{listing*}

\begin{marginlisting}
\caption{Hello world in C (minted, margin listing).}
\begin{minted}{c}
#include <stdio.h>
int main() {
  printf("Hello, World!");
  return 0;
}
\end{minted}
\end{marginlisting}

\lipsum[1]

\newpage

\begin{listing}[h]
\caption{Hello world in C (minted, normal listing).}
\begin{minted}{c}
#include <stdio.h>
int main() {
  printf("Hello, World!");
  return 0;
}
\end{minted}
\end{listing}

\lipsum[1]

\begin{listing*}[h]
\caption{Hello world in C (minted, wide listing).}
\begin{minted}{c}
#include <stdio.h>
int main() { printf("Hello, World!"); return 0; }
\end{minted}
\end{listing*}

\begin{marginlisting}
\caption{Hello world in C (minted, margin listing).}
\begin{minted}{c}
#include <stdio.h>
int main() {
  printf("Hello, World!");
  return 0;
}
\end{minted}
\end{marginlisting}

\lipsum[1]

\newpage

\begin{listing}[h]
\caption{Hello world in C (listings, normal listing).}
\begin{lstlisting}[language=C]
#include <stdio.h>
int main() {
  printf("Hello, World!");
  return 0;
}
\end{lstlisting}
\end{listing}

\lipsum[1]

\begin{listing*}[h]
\caption{Hello world in C (listings, wide listing).}
\begin{lstlisting}[language=C]
#include <stdio.h>
int main() { printf("Hello, World!"); return 0; }
\end{lstlisting}
\end{listing*}

\begin{marginlisting}
\caption{Hello world in C (listings, margin listing).}
\begin{lstlisting}[language=C]
#include <stdio.h>
int main() {
  printf("Hello, World!");
  return 0;
}
\end{lstlisting}
\end{marginlisting}

\lipsum[1]

\newpage

\begin{listing}[h]
\caption{Hello world in C (listings, normal listing).}
\begin{lstlisting}[language=C]
#include <stdio.h>
int main() {
  printf("Hello, World!");
  return 0;
}
\end{lstlisting}
\end{listing}

\lipsum[1]

\begin{listing*}[h]
\caption{Hello world in C (listings, wide listing).}
\begin{lstlisting}[language=C]
#include <stdio.h>
int main() { printf("Hello, World!"); return 0; }
\end{lstlisting}
\end{listing*}

\begin{marginlisting}
\caption{Hello world in C (listings, margin listing).}
\begin{lstlisting}[language=C]
#include <stdio.h>
int main() {
  printf("Hello, World!");
  return 0;
}
\end{lstlisting}
\end{marginlisting}

\lipsum[1]

\end{document}
```

**Expected behavior:**

[after.pdf](https://github.com/fmarotta/kaobook/files/8707576/after.pdf)

**Actual behavior:**

[before.pdf](https://github.com/fmarotta/kaobook/files/8707570/before.pdf)

### PR DESCRIPTION
**Pull request by pi8027**
This PR introduces a new class option `listing=(listings|minted)` and fixes #236. Now, `kao.sty` requires either `listings.sty` or `minted.sty` depending on the users' choice. The `marginlisting` environment supports only the chosen package.

`listings`:
- Fix the spacing issue in `marginlisting`. This is done by supplying `format=margin` to `\captionsetup`.
- *There is no wide `lstlisting` environment yet*.

`minted`:
- Add the wide `listing*` environment.
- Count `\c@listing` within chapter.

Other changes:
- The separator in captions in `marginlisting` is redone using `labelsep=colon`, *but the `numbers=enddot` class option is not properly supported*.
- The `\floatsetup` declarations are refactored.

### CODE REVIEW

**Comment by pi8027** [comment]
I found that this PR does not work with the listings package. I will try to fix it.

**Comment by fmarotta** [comment]
Hi, thanks! What's the state of this PR, after closing #235?

**Comment by pi8027** [comment]
Sorry. I still need some more time to properly redo this PR.

File: kao.sty [hunk 1]
```diff
@@ -609,7 +609,7 @@
 	\marginskip{\kaomarginskipabove}%
 	\begin{lrbox}{\@sidenotes@marginlistingbox}%
 	\begin{minipage}{\marginparwidth}%
-		\captionsetup{type=lstlisting}%
+		\captionsetup{type=listing}%
```
Comment (pi8027) [inline kao.sty:612]: It seems to me that unifying the counters for the `lstlisting` and `listing` environments is very difficult. And I cannot statically pick one of them here. Now I think I should rather introduce a new option to let the user pick `listings` or `minted` (and keep `\captionsetup*`), which requires some more refactoring work...

**Comment by pi8027** [comment]
Almost there! I still have to do something with TOC (List of Listings), which is not very clear to me at the moment. The first commit, which properly fixes the sparing issue in `marginlisting`, might be useful on its own.

**Comment by pi8027** [comment]
Done.

**Comment by ehhc** [comment]
@fmarotta any news on this? Will it be merge soon?
btw: thanks for your great work :)

**Comment by fmarotta** [comment]
Great work, thanks @pi8027! Yes, I will merge this right now.

**Comment by fmarotta** [comment]
I've added some docs and credits to the documentation: https://github.com/fmarotta/kaobook/commit/e4ee5fb3cb3d7e665eca807978040054291e074c

**Comment by ehhc** [comment]
Hi, it might be a stupid question, but if i need a minted listing* that spans above more than one page, the minted documentation (https://tug.ctan.org/macros/latex/contrib/minted/minted.pdf) states i need to create a custom envorinment (see below).

Unfortunately, if i do so, i cannot use the listing* that uses the whole document width anymore... Do you know how to solve that?

I tried to figure out what you did in this PR but, due to my little latex knowledge, was not able to figure out how you created the listing* environment at all..


> I need a listing environment that supports page breaks. The built-in listing
> environment is a standard float; it doesn’t support page breaks. You will probably
> want to define a new environment for long floats. For example,
> \usepackage{caption}
> \newenvironment{longlisting}{\captionsetup{type=listing}}{}

**Pull request merged**